src/gui.h

```diff
@@ -0,0 +1,473 @@
+     */
+    void open() {
+        ImGui::OpenPopup(_name.c_str());
+    }
+
+private:
+    std::string _name;
+    bool _active;
+};
+
+/**
+ * @brief   TODO
+ */
+class menu : public widget<menu> {
+public:
+    menu(const std::string& name):
+        _active(ImGui::BeginMenu(name.c_str()))
+    {}
+
+    ~menu() {
+        if (_active) {
+            ImGui::EndMenu();
+        }
+    }
+
+    menu(const menu&) = delete;
+    menu(menu&&) = delete;
+    menu& operator=(const menu&) = delete;
+    menu& operator=(menu&&) = delete;
+
+    /**
+     * @brief   Create a menu item with a callback
+     */
+    menu& item(const std::string& name, std::invocable auto&& callback = [] {}, bool* selected = nullptr, const std::string& shortcut = "") {
+        if (_active) {
+            if (ImGui::MenuItem(name.c_str(), shortcut == "" ? nullptr : shortcut.c_str(), selected)) {
+                std::invoke(callback);
+            }
+        }
+        return *this;
+    }
+
+private:
+    bool _active;
+};
+
+/**
+ * @brief   TODO
+ */
+class main_menu : public widget<main_menu> {
+public:
+    main_menu():
+        _active(ImGui::BeginMainMenuBar())
+    {}
+
+    ~main_menu() {
+        if (_active) {
+            ImGui::EndMainMenuBar();
+        }
+    }
+
+    main_menu(const main_menu&) = delete;
+    main_menu(main_menu&&) = delete;
+    main_menu& operator=(const main_menu&) = delete;
+    main_menu& operator=(main_menu&&) = delete;
+
+    /**
+     * @brief   TODO
+     */
+    ui::menu menu(const std::string& name) {
+        return ui::menu(name);
+    }
+
+private:
+    bool _active;
+};
+
+} // namespace ui
+
+/**
+ * @brief   Wrapper class around GLFW and ImGUI
+ */
+class gui {
+public:
+    gui(const std::string& title, vec2 size);
+    ~gui();
+
+    gui(const gui&)             = delete;
+    gui(gui&& other) noexcept;
+    gui& operator=(const gui&)  = delete;
+    gui& operator=(gui&&)       = delete;
+
+    void run(std::invocable auto&& callback);
+
+    /**
+     * @brief   TODO
+     */
+    ui::main_menu main_menu() {
+        return ui::main_menu();
+    }
+
+    /**
+     * @brief   TODO
+     */
+    ui::popup popup(const std::string& name) {
+        return ui::popup(name);
+    }
+
+    /**
+     * @brief   TODO
+     */
+    ui::window window(const std::string& title, const ui::window_config& config, bool* isOpen = nullptr) {
+        return ui::window(title, config, isOpen);
+    }
+
+    [[nodiscard]] inline auto dimensions() const noexcept { return _dimensions; }
+    [[nodiscard]] inline auto height() const noexcept { return _height; }
+    [[nodiscard]] inline auto width() const noexcept { return _width; }
+
+private:
+    std::string_view _glslVersion;
+    GLFWwindow* _window = nullptr;
+    vec2 _dimensions;
+    int _width, _height;
+
+    static void clear_color(const vec4& color);
+    static std::string_view GLSLVersion();
+};
+
+/**
+ * @brief   Event-loop
+ */
+void gui::run(std::invocable auto&& callback) {
```

Comment:
For long term: Wrapper around ImGui and around graphics API should be separated.